core/chains/evm/logpoller/parser_test.go

```diff
@@ -46,16 +47,19 @@ func TestDSLParser(t *testing.T) {
 		expressions := []query.Expression{
 			NewAddressFilter(common.HexToAddress("0x42")),
 			NewEventSigFilter(common.HexToHash("0x21")),
+			NewConfirmationsFilter(types.Finalized),
 		}
-		limiter := query.NewLimitAndSort(query.CursorLimit("10-0x42-5", query.CursorFollowing, 20))
+		limiter := query.NewLimitAndSort(query.CursorLimit("10-5-0x42", query.CursorFollowing, 20))
 
 		result, args, err := parser.buildQuery(chainID, expressions, limiter)
 		expected := "SELECT evm.logs.* " +
 			"FROM evm.logs " +
 			"WHERE evm_chain_id = :evm_chain_id " +
-			"AND (address = :address_0 AND event_sig = :event_sig_0) " +
-			"AND block_number >= :cursor_block AND tx_hash >= :cursor_txhash AND log_index > :cursor_log_index " +
-			"ORDER BY block_number ASC, tx_hash ASC, log_index ASC " +
+			"AND (address = :address_0 AND event_sig = :event_sig_0 " +
+			"AND block_number <= " +
+			"(SELECT finalized_block_number FROM evm.log_poller_blocks WHERE evm_chain_id = :evm_chain_id ORDER BY block_number DESC LIMIT 1)) " +
+			"AND block_number >= :cursor_block AND tx_hash >= :tx_hash AND log_index > :cursor_log_index " +
+			"ORDER BY block_number ASC, log_index ASC, tx_hash ASC " +
```

Comment:
It's harmless to return the `tx_hash` as a result, or to include it at the end of an ORDER BY (since the PRIMARY KEY on the table is (block_hash, log_index, evm_chain_id), which guarantees the tx_hash part of the ORDER BY will be ignored due to it being impossible for that to have any effect).  But I don't think we should include `tx_hash` in the AND clause like this, as that could return the wrong results if it did ever matter.

In principle, it shouldn't matter, since finalized logs are not supposed to ever change after they're written to the table.  But if that assumption were violated then there's a 50/50 chance you'd get the wrong result back, since the hash is entirely random.  Doesn't that defeat the only rationale for ius doing anything with the tx_hash in the first place? Instead of improving the robustness in a hypothetical edge case, it makes the results more unpredictable and error-prone.

Consider the following scenario:

1. QueryKey is called, asking for the most recent logs of a particular type in order of ascending timestamp. 
It returns 20 logs with the most recent having block_number=10, log_index=5, tx_hash=0x84.
2. A re-org happens (or backup logpoller runs, updating an old log), so that block is now replaced with a whole new set of logs. The log in the new block with log_index=4 still has the same block_number=5, but now it has a tx hash that happens to begin with 0x13.
3. QueryKey is called again, using CursorLimit("10-5-0x42", CursorFollowing, 20) to get any logs more recent than that back.
4. Because through random chance, the hash of the new transaction is less than the hash of the old one, the new one is not returned in the query.

This makes me realize something else also... why are we using >= for block_number and tx_hash but > for log_index?

Reply:
Actually, come to think of it.. there are worse problems with this query. Including tx_hash here will cause it to return the wrong results even in the typical case, not just in a hypothetical scenario where a finalized log gets updated after it's marked finalized.

Combining these range restrictions with AND doesn't work here.  The restriction `tx_hash > :tx_hash` filters out any results whose hashes don't happen to satisfy that criteria regardless of what block they are in or what their log_index is.  So even if you had, say 100 important log results in block 11, saying block_number >= 10 AND tx_hash >= 0x80 will filter out approximately 50 from the results!

And the problem is bigger than that, it's not just with tx_hash the same problem can happen with log_index.  If the starting point of your cursor has a high log_index, this will never give you any results with a lower log_index even if they are in a more recent block.

Reply:
However it gets built, we need the query to end up looking like this:
```
block_number > :block_number OR (block_number = :block_number AND log_index > :log_index)
```

Reply:
I updated the query to match the `OR` condition you created above.